Card: migrate to TypeScript

torounit opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of https://github.com/WordPress/gutenberg/issues/35744

Converts the `Card` component to TypeScript.

## Testing Instructions

1. `npm run storybook:dev`
2. check `Card`

torounit committed:
rename to ts

torounit committed:
Replace component with tsx.

torounit committed:
move MediaProps to types.ts

torounit committed:
fix imports

torounit committed:
component props to DRY

torounit committed:
convert to ts.

torounit committed:
fix type.

torounit committed:
remove generics

torounit committed:
fix types

torounit committed:
named export

torounit committed:
remove args

ciampo commented on packages/components/src/card/card-header/component.tsx:
```diff
@@ -1,15 +1,19 @@
 /**
- * @param {import('../../ui/context').WordPressComponentProps<import('../types').HeaderProps, 'div'>} props
- * @param {import('react').ForwardedRef<any>}                                                         forwardedRef
+ * External dependencies
```
External dependencies should be listed before internal dependencies

ciampo commented on packages/components/src/card/card/hook.ts:
```diff
@@ -7,17 +7,20 @@ import { useMemo } from '@wordpress/element';
+	isElevated,
+	...otherProps
+}: CardProps ): CardProps {
+	const propsToReturn: CardProps = {
```
I believe the most correct way to code this hook is 

```suggestion
}: CardProps ) {
	const propsToReturn: Omit< CardProps, 'isElevated' > = {
```

1. Let's not add an explicit return value for the function, and instead rely on TypeScript inferring it
2. Let's be explicit about the fact that `isElevated` will be excluded from the return value of this hook

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -156,7 +156,7 @@ function Example(
 A couple of good examples of how hooks are used for composition are:
 
-- the `Card` component, which builds on top of the `Surface` component by [calling the `useSurface` hook inside its own hook](/packages/components/src/card/card/hook.js);
+- the `Card` component, which builds on top of the `Surface` component by [calling the `useSurface` hook inside its own hook](/packages/components/src/card/card/hook.ts);
```
It looks like there are more references left to `Card` files ending in `*.js` in this file, that also need updating

ciampo reviewed (commented):
Hey @torounit , thank you for working on this!

On top of the inline comments, here's some additional feedback:

- Should we add a wrapper around the `Card` in the Storybook example?
- Currently, Storybook is not showing controls and documentation for the subcomponents — how can we fix that ? (cc @mirka )
- Can we also refactor the unit tests to TypeScript?

Thank you!

mirka commented:
> * Currently, Storybook is not showing controls and documentation for the subcomponents — how can we fix that ? (cc @mirka )

Hmm, the subcomponents are showing for me. Maybe yours needed a cold restart?

https://user-images.githubusercontent.com/555336/183650847-343ca5a7-14ef-4761-9d4c-f6cf84f72339.mp4

(Subcomponents don't support controls, btw, so only the docs will show)

ciampo commented:
Ah! It took me a while to find them!

https://user-images.githubusercontent.com/1083581/183933426-0d1d335b-4e37-4567-9dea-1f9b5553067b.mp4

> (Subcomponents don't support controls, btw, so only the docs will show)

That's unfortunate. Hopefully support will be added soon

torounit committed:
fix CONTRIBUTING.md

torounit committed:
fix import order

torounit committed:
Update packages/components/src/card/card/hook.ts

Co-authored-by: Marco Ciampini <marco.ciampo@gmail.com>

torounit committed:
convert to typescript

torounit committed:
remove `@example` .

torounit committed:
add wrapper

torounit commented:
@ciampo 

Sorry for the late reply, thanks for the review.

* Added Wrapper to the story.
* Fixed CONTRIBUTING.md.
* Refactored unit tests to TypeScript.

ciampo commented on packages/components/CONTRIBUTING.md:
```diff
@@ -276,7 +276,7 @@ export function useCard( props ) {
 
 //=========================================================================
-// Simplified snippet from `packages/components/src/card/card/component.js`
+// Simplified snippet from `packages/components/src/card/card/component.ts`
```
I found one more instance that needs renaming:

https://github.com/WordPress/gutenberg/blob/8f82fbc970a4250fee4dd7ac333aa890f69af16d/packages/components/CONTRIBUTING.md?plain=1#L316

ciampo commented on packages/components/src/card/test/index.tsx:
```diff
@@ -56,9 +56,9 @@ describe( 'Card', () => {
-			);
+			expect(
+				containerRounded.firstElementChild
+			).toMatchStyleDiffSnapshot( containerSquared.firstElementChild );
```
We also need to rename `packages/components/src/card/test/__snapshots__/index.js.snap` to `packages/components/src/card/test/__snapshots__/index.tsx.snap`

ciampo reviewed (commented):
We're almost there! Just left a couple of minor comments.

Thank you for your patience!

torounit committed:
fix path

torounit committed:
rename snapshot

ciampo commented:
This PR may need a rebase to pass CI tests

ciampo commented:
> This PR may need a rebase to pass CI tests

Ah, and a CHANGELOG entry (the "internal" section is ok)

torounit committed:
add CHANGELOG.md

ciampo committed:
Merge branch 'trunk' into refactor/ts-card
